onnxruntime_extensions/tools/pre_post_processing/steps/nlp.py

```diff
@@ -0,0 +1,449 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import onnx
+from collections import OrderedDict
+
+from typing import List, Optional, Tuple, Union
+from ..step import Step
+
+
+class SentencePieceTokenizer(Step):
+    def __init__(self, tokenizer, domain: str = "ai.onnx.contrib", name: Optional[str] = None):
+        """
+        Args:
+            tokenizer: SentencePiece tokenizer from huggingface,
+                usually, we can get it by "tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)"
+            domain: Optional domain of step. ["ai.onnx.contrib" or "com.microsoft.extensions"] Defaults to 'ai.onnx.contrib'
+            name: Optional name of step. Defaults to 'SentencePieceTokenizer'
+
+        """
+        super().__init__(
+            ["inputs", "nbest_size", "alpha", "add_bos", "add_eos", "reverse"], ["input_ids", "attention_mask"], name
+        )
+        self.domain = domain
+        assert domain in [
+            "ai.onnx.contrib",
+            "com.microsoft.extensions",
+        ], f"invalid domain:`{domain}` for SentencePieceTokenizer"
+        self.tokenizer = tokenizer
+
+    def _create_graph_for_step(self, graph: onnx.GraphProto, onnx_opset: int):
+        graph_input_names = [inp.name for inp in graph.output]
+        assert self.input_names[0] in graph_input_names
+
+        # input text
+        input_type_str0, input_shape_str0 = self._get_input_type_and_shape_strs(graph, 0)
+
+        input_type_str1, input_shape_str1 = (
+            self._get_input_type_and_shape_strs(graph, 1)
+            if self.input_names[1] in graph_input_names
+            else ("int64", "0")
+        )
+        input_type_str2, input_shape_str2 = (
+            self._get_input_type_and_shape_strs(graph, 2)
+            if self.input_names[2] in graph_input_names
+            else ("float", "0")
+        )
+        input_type_str3, input_shape_str3 = (
+            self._get_input_type_and_shape_strs(graph, 3) if self.input_names[3] in graph_input_names else ("bool", "0")
+        )
+        input_type_str4, input_shape_str4 = (
+            self._get_input_type_and_shape_strs(graph, 4) if self.input_names[4] in graph_input_names else ("bool", "0")
+        )
+        input_type_str5, input_shape_str5 = (
+            self._get_input_type_and_shape_strs(graph, 5) if self.input_names[5] in graph_input_names else ("bool", "0")
+        )
```

Comment:
Not quite following all the complexity here. Before each Step is run the PrePostProcessor connects things up with the producers of the input values which includes updating the input_names.

What's the reason for all the additional logic checking individual inputs? I don't recall having to do that for any of the other steps.

Reply:
Can we have a flag in the ctor which chooses between a single input or 5 and branch on that in the code?

Reply:
When we create inputs for `PrePostProcessor`, the function `create_named_value` required uers given a name for the input_value_info, for example
```
create_named_value("text",.....) # required
create_named_value("nbest_size",.....) #optional 
create_named_value("add_bos",.....) #optional 
create_named_value("add_eos",.....) #optional 
create_named_value("reverse",.....) #optional 
```

Is it enough for the purpose?

Reply:
Sounds good. Thanks.